File: project_euler/Problem030/Problem030.java
package project_euler;

public class Problem030 {
    public void solution() {
        int target = 5;
        int[] powers = new int[10];
        for (int i = 0; i < 10; ++i) {
            powers[i] = (int) Math.pow(i, target);
        }

        int result = 0;
        for (int i = 6 * powers[9]; i >= 10; --i) {
            if (i == getPowerDigit(i, powers)) {
                result += i;
            }
        }
        System.out.println(result);
    }

    private int getPowerDigit(int value, int[] powers) {
        int result = 0;
        while (0 != value) {
            result += powers[value % 10];
            value /= 10;
        }
        return result;
    }
}

File: project_euler/Problem030/Problem030.go
package project_euler

import (
    "fmt"
    "math"
)

func Problem030() {
    target := 5
    powers := make([]int, 10)
    for i := 0; i < 10; i++ {
        powers[i] = int(math.Pow(float64(i), float64(target)))
    }

    result := 0
    for i := 6 * powers[9]; i >= 10; i-- {
        if i == getPowerDigit(i, powers) {
            result += i
        }
    }
    fmt.Println(result)
}

func getPowerDigit(value int, powers []int) int {
    result := 0
    for 0 != value {
        result += powers[value % 10]
        value /= 10
    }
    return result
}
